File: C/1/19.c
#include <stdio.h>

#define MAXLINE 1000   /* maximum input line length */

int getline(char line[], int maxline);
void reverse(char line[], char reversed[], int len);

/* Remove trailing blanks and whitespaces, delete blank lines. */
main()
{
    int len, idx;    /* current line length */
    char line[MAXLINE];    /* current input line */
    char reversed[MAXLINE];    /* reversed input line */

    while ((len = getline(line, MAXLINE)) > 0) {
        reverse(line, reversed, len);

        printf("READ: %04d; | %s\n", len, reversed);

        /* We need to clear out the reversed string. */
        for (idx = 0; idx < MAXLINE; idx++) {
            reversed[idx] = '\0';
        }
    }

    return 0;
}

/* getline:  read a line into s, return length  */
int getline(char s[],int lim)
{
    int c, i, last = 0;

    for (i=0; i < lim-1 && (c=getchar())!=EOF && c!='\n'; ++i)
        s[i] = c;

    /* Because of operator precedence, we don't need to check if i < lim -1 */
    if (c == '\n') {
        s[i] = c;
        ++i;
    }

    /* Adding the null character */
    s[i] = '\0';

    return i;
}

/**
 * Removes trailing blanks and tabs from a line
 * 
 * @param char line[]       String from input
 * @param char reversed[]   Array with result
 * @param int  len          String size
 * 
 * @return void
 */
void reverse(char line[], char reversed[], int len)
{
    int i = 0, j = 0;

    for (i = len-2; i >=0; i--, j++) {
        reversed[j] = line[i];
    }

    reversed[len] = '\0';
}

File: rustlang/1/19.rs
use std::io;

/// Remove trailing blanks and whitespaces, delete blank lines.
fn main ()
{
    // a mutable variable to hold input data
    let mut input = String::new();

    // using rust infinite loop
    loop {
        match io::stdin().read_line(&mut input) {
            // io::Result::Ok receives the number of bytes read
            Ok (n_bytes) => {
                // case EOF is reached, "read_lines" will return 0 and we break out of the loop
                if n_bytes == 0 {
                    break;
                }

                // In Rust it's a bit easier to do...
                let reversed = reverse(&input);
                println!("READ: {:04}; | {}", input.chars().count(), reversed);

                // we need to clear input buffer because "read_line" appends all read bytes to it
                input.clear();
            },
            Err(_) => {
                panic!("Unexpected error.");
            }
        }
    }
}

/// A function that borrows the original input and return it reversed.
fn reverse (original : &String) -> String {
    // to match C code we remove trailing newlines
    // also, we use trait methods to reverse char order and collect everything
    // into a new String object
    original.trim_end_matches('\n').chars().rev().collect::<String>()
}
